# Lab 02: Basic NLP Preprocessing Techniques

**Course:** ITAI 2373 - Natural Language Processing  
**Module:** 02 - Text Preprocessing  
**Duration:** 2-3 hours  
**Student Name:** ___Troy Nsofor_____________  
**Date:** __9/20/2025______________

---

## 🎯 Learning Objectives

By completing this lab, you will:
1. Understand the critical role of preprocessing in NLP pipelines
2. Master fundamental text preprocessing techniques
3. Compare different libraries and their approaches
4. Analyze the effects of preprocessing on text data
5. Build a complete preprocessing pipeline
6. Load and work with different types of text datasets

## 📖 Introduction to NLP Preprocessing

Natural Language Processing (NLP) preprocessing refers to the initial steps taken to clean and transform raw text data into a format that's more suitable for analysis by machine learning algorithms.

### Why is preprocessing crucial?

1. **Standardization:** Ensures consistent text format across your dataset
2. **Noise Reduction:** Removes irrelevant information that could confuse algorithms
3. **Complexity Reduction:** Simplifies text to focus on meaningful patterns
4. **Performance Enhancement:** Improves the efficiency and accuracy of downstream tasks

### Real-world Impact
Consider searching for "running shoes" vs "Running Shoes!" - without preprocessing, these might be treated as completely different queries. Preprocessing ensures they're recognized as equivalent.

### 🤔 Conceptual Question 1
**Before we start coding, think about your daily interactions with text processing systems (search engines, chatbots, translation apps). What challenges do you think these systems face when processing human language? List at least 3 specific challenges and explain why each is problematic.**

*Double-click this cell to write your answer:*

**Challenge 1:** The first challenge that they have is that they have difficulty with speech recognition. Reasons why the speech recognition is problematic are that there are accents and dialects. There are background sounds, homophones and words that sound the same. There are speech impediments and pace.

**Challenge 2:** The second challenge that they have a problem with is word recognition. The reasons why are because of amiguity and polysemy. There is contextual learniing, slang, idioms, and colloquilisms. There are misspellings and grammatical mistakes. There is also not enoguh of feelings and empathy.

**Challenge 3:** The third challenge that they have is with language. The same reasons from challenge #2 are why it is problematic.

---

## 🛠️ Part 1: Environment Setup

We'll be working with two major NLP libraries:
- **NLTK (Natural Language Toolkit):** Comprehensive NLP library with extensive resources
- **spaCy:** Industrial-strength NLP with pre-trained models

**⚠️ Note:** Installation might take 2-3 minutes to complete.

In [1]:
# Step 1: Install Required Libraries
print("🔧 Installing NLP libraries...")

!pip install -q nltk spacy
!python -m spacy download en_core_web_sm

print("✅ Installation complete!")

🔧 Installing NLP libraries...
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 12.8/12.8 MB 84.7 MB/s eta 0:00:00
✔ Download and installation successful
You can now load the package via spacy.load('en_core_web_sm')
⚠ Restart to reload dependencies
If you are in a Jupyter or Colab notebook, you may need to restart Python in
order to load all the package's dependencies. You can do this by selecting the
'Restart kernel' or 'Restart runtime' option.
✅ Installation complete!


In [2]:
'en_core_web_sm'

'en_core_web_sm'

### 🤔 Conceptual Question 2
**Why do you think we need to install a separate language model (en_core_web_sm) for spaCy? What components might this model contain that help with text processing? Think about what information a computer needs to understand English text.**

*Double-click this cell to write your answer:* We need to install a separate language model like en_core_web_sm for spaCy because the main spaCy library itself doesn't really "understand" English-it just has the equipment to examine text. The language model adds the intelligence that allows the computer to learn words, sentences, and grammar. The model has things like vocabulary, grammar rules, and patterns that help spaCy figure out things such as what words are the nouns or verbs, how words refer to each other in a sentence, and also to know name of people, places, companies, etc. Additionally it gives the computer the details it requires to make English text make sense. If there's no language model, spaCy would'nt do more of the text processing tasks we want or need it to do.


---

In [3]:
# Step 2: Import Libraries and Download NLTK Data
import nltk
import spacy
import string
import re
from collections import Counter

# Download essential NLTK data
print("📦 Downloading NLTK data packages...")
nltk.download('punkt')      # For tokenization
nltk.download('stopwords')  # For stop word removal
nltk.download('wordnet')    # For lemmatization
nltk.download('averaged_perceptron_tagger')  # For POS tagging

print("\n✅ All imports and downloads completed!")

📦 Downloading NLTK data packages...


[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.



✅ All imports and downloads completed!


[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.
[nltk_data] Downloading package wordnet to /root/nltk_data...
[nltk_data] Downloading package averaged_perceptron_tagger to
[nltk_data]     /root/nltk_data...
[nltk_data]   Unzipping taggers/averaged_perceptron_tagger.zip.


In [4]:
('punkt')

'punkt'

In [5]:
('stopwords')

'stopwords'

In [6]:
('wordnet')

'wordnet'

In [7]:
('averaged_perceptron_tagger')

'averaged_perceptron_tagger'

In [8]:
("Downloading NLTK data packages...")

'Downloading NLTK data packages...'

## 📂 Part 2: Sample Text Data

We'll work with different types of text to understand how preprocessing affects various text styles:
- Simple text
- Academic text (with citations, URLs)
- Social media text (with emojis, hashtags)
- News text (formal writing)
- Product reviews (informal, ratings)

In [9]:
# Step 3: Load Sample Texts
simple_text = "Natural Language Processing is a fascinating field of AI. It's amazing!"

academic_text = """
Dr. Smith's research on machine-learning algorithms is groundbreaking!
She published 3 papers in 2023, focusing on deep neural networks (DNNs).
The results were amazing - accuracy improved by 15.7%!
"This is revolutionary," said Prof. Johnson.
Visit https://example.com for more info. #NLP #AI @university
"""

social_text = "OMG! Just tried the new coffee shop ☕️ SO GOOD!!! Highly recommend 👍 #coffee #yum 😍"

news_text = """
The stock market experienced significant volatility today, with tech stocks leading the decline.
Apple Inc. (AAPL) dropped 3.2%, while Microsoft Corp. fell 2.8%.
"We're seeing a rotation out of growth stocks," said analyst Jane Doe from XYZ Capital.
"""

review_text = """
This laptop is absolutely fantastic! I've been using it for 6 months and it's still super fast.
The battery life is incredible - lasts 8-10 hours easily.
Only complaint: the keyboard could be better. Overall rating: 4.5/5 stars.
"""

# Store all texts
sample_texts = {
    "Simple": simple_text,
    "Academic": academic_text.strip(),
    "Social Media": social_text,
    "News": news_text.strip(),
    "Product Review": review_text.strip()
}

print("📄 Sample texts loaded successfully!")
for name, text in sample_texts.items():
    preview = text[:80] + "..." if len(text) > 80 else text
    print(f"\n🏷️ {name}: {preview}")

📄 Sample texts loaded successfully!

🏷️ Simple: Natural Language Processing is a fascinating field of AI. It's amazing!

🏷️ Academic: Dr. Smith's research on machine-learning algorithms is groundbreaking!
She publi...

🏷️ Social Media: OMG! Just tried the new coffee shop ☕️ SO GOOD!!! Highly recommend 👍 #coffee #yu...

🏷️ News: The stock market experienced significant volatility today, with tech stocks lead...

🏷️ Product Review: This laptop is absolutely fantastic! I've been using it for 6 months and it's st...


In [10]:
('simple_text')

'simple_text'

In [11]:
('academic_text.strip')

'academic_text.strip'

In [12]:
('social_text')

'social_text'

In [13]:
('news_text.strip')

'news_text.strip'

In [14]:
('review_text.strip')

'review_text.strip'

In [15]:
("Sample texts loaded successfully!")

'Sample texts loaded successfully!'

In [16]:
(f"\n {'name'}: {'preview'}")

'\n name: preview'

### 🤔 Conceptual Question 3
**Looking at the different text types we've loaded, what preprocessing challenges do you anticipate for each type? For each text type below, identify at least 2 specific preprocessing challenges and explain why they might be problematic for NLP analysis.**

*Double-click this cell to write your answer:*

**Simple text challenges:**
1. Short length & limited context - Simple sentences usually don't provide a lot of surronding text, so NLP models may not have enough context for correct classification or subject modeling
2. Subjectivity markers ("amazing") - Sentiment words are undetermined without intesity scaling; they can turn sentiment examinations if they're not formalized or interpreted accuretly.

**Academic text challenges:**
1. Techinical vocabulary & domain-specific jargon - Needs specialized tokenization and domain adaptation, otherwise models may fail to catch the definition.
2. Abbreviations & citations - Words such as et al., Fig. 2, IEEE or shorten words ("publ...") might confuse standard preprocessing and trigger sounds.

**Social media text challenges:**
1. Non-standard language (slang, emojis, acronyms, elongated words) - Preprocessing needs to nomarlize "SO GOOD!!!" vs. "good" or "OMG" vs. "of my god."
2. Hastags, mentions, links - The tokens might carry a definition (#coffee) or be unrelated (spam hastags), making it challenging whether to keep or delete them.

**News text challenges:**
1. Named entities & proper nouns - Spotting businesses, markets, people need reliable Named Entity Recognition (NER)
2. Formal but shortened text ("lead...") - Separating or breaking down sentences (common in headlines or snippets) makes parsing difficult and might distort context.

**Product review challenges:**
1. Subjective language & exaggeration - Words like "fantastic" or "absoltutely" might exaggerate sentiment, needing careful handling in sentiment scoring.
2. Aspect-based sentiment - Multiple product parts (e.g., battery, screen, performance) might be hidden in one review, needing finer-grained extraction than easy positive/negative classification.

---

## 🔤 Part 3: Tokenization

### What is Tokenization?
Tokenization is the process of breaking down text into smaller, meaningful units called **tokens**. These tokens are typically words, but can also be sentences, characters, or subwords.

### Why is it Important?
- Most NLP algorithms work with individual tokens, not entire texts
- It's the foundation for all subsequent preprocessing steps
- Different tokenization strategies can significantly impact results

### Common Challenges:
- **Contractions:** "don't" → "do" + "n't" or "don't"?
- **Punctuation:** Keep with words or separate?
- **Special characters:** How to handle @, #, URLs?

In [33]:
# Step 4: Tokenization with NLTK
from nltk.tokenize import word_tokenize, sent_tokenize

# Test on simple text
print("🔍 NLTK Tokenization Results")
print("=" * 40)
print(f"Original: {simple_text}")

# Word tokenization
nltk_tokens = word_tokenize(simple_text)
print(f"\nWord tokens: {nltk_tokens}")
print(f"Number of tokens: {len(nltk_tokens)}")

# Sentence tokenization
sentences = sent_tokenize(simple_text)
print(f"\nSentences: {sentences}")
print(f"Number of sentences: {len(sentences)}")

🔍 NLTK Tokenization Results
Original: Natural Language Processing is a fascinating field of AI. It's amazing!

Word tokens: ['Natural', 'Language', 'Processing', 'is', 'a', 'fascinating', 'field', 'of', 'AI', '.', 'It', "'s", 'amazing', '!']
Number of tokens: 14

Sentences: ['Natural Language Processing is a fascinating field of AI.', "It's amazing!"]
Number of sentences: 2


In [18]:
nltk_tokens = 'word_tokenize'

In [19]:
sentences = 'sent_tokenize'

In [20]:
'sent_tokenize'

'sent_tokenize'

In [21]:
'word_tokenize'

'word_tokenize'

In [22]:
'nltk_tokenize'

'nltk_tokenize'

In [23]:
("🔍 NLTK Tokenization Results")

'🔍 NLTK Tokenization Results'

In [24]:
("=" * 40)

'========================================'

In [25]:
( f"Original: {simple_text}")

"Original: Natural Language Processing is a fascinating field of AI. It's amazing!"

In [26]:
(f"\nWord tokens: {'nltk_tokens'}")

'\nWord tokens: nltk_tokens'

In [27]:
(f"Number of tokens: {len('nltk_tokens')}")

'Number of tokens: 11'

In [28]:
(f"\nSentences: {'sentences'}")

'\nSentences: sentences'

In [29]:
(f"Number of sentences: {len('sentences')}")

'Number of sentences: 9'

In [30]:
('= * 40')

'= * 40'

In [31]:
nltk.download('punkt_tab')

[nltk_data] Downloading package punkt_tab to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt_tab.zip.


True

In [32]:
('punkt_tab')

'punkt_tab'

### 🤔 Conceptual Question 4
**Examine the NLTK tokenization results above. How did NLTK handle the contraction "It's"? What happened to the punctuation marks? Do you think this approach is appropriate for all NLP tasks? Explain your reasoning.**

*Double-click this cell to write your answer:*

**How "It's" was handled:** The NLTK handled the contraction "It's" by putting it in sentences. It also separated the letter (s) from the word (It's) in the Word Tokens section.

**Punctuation treatment:** The punctation marks in the Word Tokens section were between the words, periods, letters, and exclaimation marks. The punctation marks were between the sentence that's in the bracket from the Sentence section.

**Appropriateness for different tasks:** NLTK's handling of the contraction "It's by splitting it into "It" and "'s" is useful for duties such as POS tagging and parsing, since it makes the grammatical structure easier to understand. But, the approach isn't always appropraite for all NLP tasks-for instance, in sentiment or topic modeling, keeping "It's" as a single token can better conserve the meaning of the word.

---

In [34]:
# Step 5: Tokenization with spaCy
nlp = spacy.load('en_core_web_sm')

print("🔍 spaCy Tokenization Results")
print("=" * 40)
print(f"Original: {simple_text}")

# Process with spaCy
doc = nlp(simple_text)

# Extract tokens
spacy_tokens = [token.text for token in doc]
print(f"\nWord tokens: {spacy_tokens}")
print(f"Number of tokens: {len(spacy_tokens)}")

# Show detailed token information
print(f"\n🔬 Detailed Token Analysis:")
print(f"{'Token':<12} {'POS':<8} {'Lemma':<12} {'Is Alpha':<8} {'Is Stop':<8}")
print("-" * 50)
for token in doc:
    print(f"{token.text:<12} {token.pos_:<8} {token.lemma_:<12} {token.is_alpha:<8} {token.is_stop:<8}")

🔍 spaCy Tokenization Results
Original: Natural Language Processing is a fascinating field of AI. It's amazing!

Word tokens: ['Natural', 'Language', 'Processing', 'is', 'a', 'fascinating', 'field', 'of', 'AI', '.', 'It', "'s", 'amazing', '!']
Number of tokens: 14

🔬 Detailed Token Analysis:
Token        POS      Lemma        Is Alpha Is Stop 
--------------------------------------------------
Natural      PROPN    Natural      1        0       
Language     PROPN    Language     1        0       
Processing   NOUN     processing   1        0       
is           AUX      be           1        1       
a            DET      a            1        1       
fascinating  ADJ      fascinating  1        0       
field        NOUN     field        1        0       
of           ADP      of           1        1       
AI           PROPN    AI           1        0       
.            PUNCT    .            0        0       
It           PRON     it           1        1       
's           AUX     

In [35]:
(f"{'token.text:,<12'} {'token.pos_:<8'} {'token.lemma_:<12'} {'token.is_alpha:<8'} {'token.is_stop:<8'}")

'token.text:,<12 token.pos_:<8 token.lemma_:<12 token.is_alpha:<8 token.is_stop:<8'

In [36]:
("-" * 50)

'--------------------------------------------------'

In [37]:
(f"{'Token':<12} {'POS':<8} {'Lemma':<12} {'Is Alpha':<8} {'Is Stop':<8}")

'Token        POS      Lemma        Is Alpha Is Stop '

In [38]:
(f"\n Detailed Token Analysis")

'\n Detailed Token Analysis'

In [39]:
(f"Number of tokens: {len('spacy_tokens')}")

'Number of tokens: 12'

In [40]:
(f"\nWord tokens: {'spacy_tokens'}")

'\nWord tokens: spacy_tokens'

In [41]:
['token.text for token in doc']

['token.text for token in doc']

In [42]:
("🔍 spaCy Tokenization Results")

'🔍 spaCy Tokenization Results'

In [43]:
nlp = spacy.load('en_core_web_sm')

In [44]:
("=" * 40)

'========================================'

In [45]:
(f"Original: {simple_text}")

"Original: Natural Language Processing is a fascinating field of AI. It's amazing!"

### 🤔 Conceptual Question 5
**Compare the NLTK and spaCy tokenization results. What differences do you notice? Which approach do you think would be better for different NLP tasks? Consider specific examples like sentiment analysis vs. information extraction.**

*Double-click this cell to write your answer:*

**Key differences observed:** The main difference is that NLTK just separates text into tokens, while spaCy not just tokenizes but even includes extra details like POS tags, lemmas, and stopword detection. For example, spaCy highlights that "'s" is an auxiliary verb with the lemma "be," while NLTK only uses it as a token. This makes NLTK easier and lightweight, but spaCy is more detailed and better for tasks that require further lingustic examination.

**Better for sentiment analysis:** NTLK's simpler tokenization is often plenty, since sentiment particulary depends on surface words (e.g., "amazing," "fascinating"). Separating contractions is okay, but informative POS tags or lemmas might not be strictly mandatory. NLTK might be more lightweight and effective here.

**Better for information extraction:** spaCy's richer annotations (POS tags, lemmas, stopwword detection, depndecy parsing) are lot more useful. Instance: drawing out subject-verb-object relationships ("Processing is fascinating") or spotting entities such as "AI". spaCy is clearly powerful for the structured duties.

**Overall assessment:** Overall, NLTK is lightweight and efficent for simple tasks that concentrate on surface-level word use, while spaCy introdcues a more comprehensive linguistic examiantion that provides advanced tasks needing structure and context. The best decision depends on the goal: use NLTK for fast, straightforward preprocessing and spaCy for further, more informative NLP applications.

---

In [46]:
# Step 6: Test Tokenization on Complex Text
print("🧪 Testing on Social Media Text")
print("=" * 40)
print(f"Original: {social_text}")

# NLTK approach
social_nltk_tokens = word_tokenize(social_text)
print(f"\nNLTK tokens: {social_nltk_tokens}")

# spaCy approach
social_doc = nlp(social_text)
social_spacy_tokens = [token.text for token in social_doc]
print(f"spaCy tokens: {social_spacy_tokens}")

print(f"\n📊 Comparison:")
print(f"NLTK token count: {len(social_nltk_tokens)}")
print(f"spaCy token count: {len(social_spacy_tokens)}")

🧪 Testing on Social Media Text
Original: OMG! Just tried the new coffee shop ☕️ SO GOOD!!! Highly recommend 👍 #coffee #yum 😍

NLTK tokens: ['OMG', '!', 'Just', 'tried', 'the', 'new', 'coffee', 'shop', '☕️', 'SO', 'GOOD', '!', '!', '!', 'Highly', 'recommend', '👍', '#', 'coffee', '#', 'yum', '😍']
spaCy tokens: ['OMG', '!', 'Just', 'tried', 'the', 'new', 'coffee', 'shop', '☕', '️', 'SO', 'GOOD', '!', '!', '!', 'Highly', 'recommend', '👍', '#', 'coffee', '#', 'yum', '😍']

📊 Comparison:
NLTK token count: 22
spaCy token count: 23


In [47]:
social_spacy_tokens = ['token.text for token in social_doc']

In [48]:
(f"spaCy token count: {len(social_spacy_tokens)}")

'spaCy token count: 1'

In [49]:
(f"NLTK token count: {len('social_nltk_tokens')}")

'NLTK token count: 18'

In [50]:
(f"\n Comparison:")

'\n Comparison:'

In [51]:
(f"spaCy tokens: {'social_spacy_tokens'}")

'spaCy tokens: social_spacy_tokens'

In [52]:
social_doc = nlp(social_text)

In [53]:
nlp(social_text)

OMG! Just tried the new coffee shop ☕️ SO GOOD!!! Highly recommend 👍 #coffee #yum 😍

In [54]:
(f"\nNLTK tokens: {'social_nltk_tokens'}")

'\nNLTK tokens: social_nltk_tokens'

In [55]:
word_tokenize(social_text)

['OMG',
 '!',
 'Just',
 'tried',
 'the',
 'new',
 'coffee',
 'shop',
 '☕️',
 'SO',
 'GOOD',
 '!',
 '!',
 '!',
 'Highly',
 'recommend',
 '👍',
 '#',
 'coffee',
 '#',
 'yum',
 '😍']

In [56]:
("=" * 40)

'========================================'

In [57]:
("🧪 Testing on Social Media Text")

'🧪 Testing on Social Media Text'

In [58]:
(f"Original: {social_text}")

'Original: OMG! Just tried the new coffee shop ☕️ SO GOOD!!! Highly recommend 👍 #coffee #yum 😍'

### 🤔 Conceptual Question 6
**Looking at how the libraries handled social media text (emojis, hashtags), which library seems more robust for handling "messy" real-world text? What specific advantages do you notice? How might this impact a real-world application like social media sentiment analysis?**

*Double-click this cell to write your answer:*

**More robust library:** The spaCy library is more robust for messy, real-world social media text. It splits elements such as the hastags, emojis, and repeated punctuation in a strategy that gives finer control. This granlarity lets me to later choose wheter to recombine tokens or treat them as signals.

**Specific advantages:** spaCy handles social media text more robustly than NLTK by keeping emojis and hashtags unbroken and accessing informal language normally. This makes it better suited for applications such as sentiment analysis, where securing the tokens is crucial.

**Impact on sentiment analysis:** Emojis and repeated punctuation particularly carry powerful emotional cues. spaCy's tokenization makes it straightfoward to catch the signals. Separating hashtags (#coffee - "coffee") helps models use them as original words while continuing to let me to track the hashtag marker. Overall, spaCy's approach can enhance correction in social media sentiment analysis by maintaining emotional intensity and meaningful symbols, whereas NLTK's simpler method risks losing variation.

---

## 🛑 Part 4: Stop Words Removal

### What are Stop Words?
Stop words are common words that appear frequently in a language but typically don't carry much meaningful information about the content. Examples include "the", "is", "at", "which", "on", etc.

### Why Remove Stop Words?
1. **Reduce noise** in the data
2. **Improve efficiency** by reducing vocabulary size
3. **Focus on content words** that carry semantic meaning

### When NOT to Remove Stop Words?
- **Sentiment analysis:** "not good" vs "good" - the "not" is crucial!
- **Question answering:** "What is the capital?" - "what" and "is" provide context

In [59]:
# Step 7: Explore Stop Words Lists
from nltk.corpus import stopwords

# Get NLTK English stop words
nltk_stopwords = set(stopwords.words('english'))
print(f"📊 NLTK has {len(nltk_stopwords)} English stop words")
print(f"First 20: {sorted(list(nltk_stopwords))[:20]}")

# Get spaCy stop words
spacy_stopwords = nlp.Defaults.stop_words
print(f"\n📊 spaCy has {len(spacy_stopwords)} English stop words")
print(f"First 20: {sorted(list(spacy_stopwords))[:20]}")

# Compare the lists
common_stopwords = nltk_stopwords.intersection(spacy_stopwords)
nltk_only = nltk_stopwords - spacy_stopwords
spacy_only = spacy_stopwords - nltk_stopwords

print(f"\n🔍 Comparison:")
print(f"Common stop words: {len(common_stopwords)}")
print(f"Only in NLTK: {len(nltk_only)} - Examples: {sorted(list(nltk_only))[:5]}")
print(f"Only in spaCy: {len(spacy_only)} - Examples: {sorted(list(spacy_only))[:5]}")

📊 NLTK has 198 English stop words
First 20: ['a', 'about', 'above', 'after', 'again', 'against', 'ain', 'all', 'am', 'an', 'and', 'any', 'are', 'aren', "aren't", 'as', 'at', 'be', 'because', 'been']

📊 spaCy has 326 English stop words
First 20: ["'d", "'ll", "'m", "'re", "'s", "'ve", 'a', 'about', 'above', 'across', 'after', 'afterwards', 'again', 'against', 'all', 'almost', 'alone', 'along', 'already', 'also']

🔍 Comparison:
Common stop words: 123
Only in NLTK: 75 - Examples: ['ain', 'aren', "aren't", 'couldn', "couldn't"]
Only in spaCy: 203 - Examples: ["'d", "'ll", "'m", "'re", "'s"]


In [60]:
common_stopwords = ('spacy_stopwords')

In [61]:
common_stopwords = 'nltk_stopwords'

In [62]:
(f"First 20: {sorted(list(spacy_stopwords))[:20]}")

'First 20: ["\'d", "\'ll", "\'m", "\'re", "\'s", "\'ve", \'a\', \'about\', \'above\', \'across\', \'after\', \'afterwards\', \'again\', \'against\', \'all\', \'almost\', \'alone\', \'along\', \'already\', \'also\']'

In [63]:
common_stopwords = 'intersection'

In [64]:
(f"Only in spaCy: {len(spacy_only)} - Examples: {sorted(list(spacy_only))[:5]}")

'Only in spaCy: 203 - Examples: ["\'d", "\'ll", "\'m", "\'re", "\'s"]'

In [65]:
(f"Only in NLTK: {len(nltk_only)} - Examples: {sorted(list(nltk_only))[:5]}")

'Only in NLTK: 75 - Examples: [\'ain\', \'aren\', "aren\'t", \'couldn\', "couldn\'t"]'

In [66]:
(f"Common stop words: {len('common_stopwords')}")

'Common stop words: 16'

In [67]:
(f"🔍 Comparison:")

'🔍 Comparison:'

In [68]:
spacy_only = 'spacy_stopwords - nltk_stopwords'

In [69]:
nltk_only = 'nltk_stopwords - spacy_stopwords'

In [70]:
(f"📊 NLTK has {len('nltk_stopwords')} English stop words")

'📊 NLTK has 14 English stop words'

In [71]:
(f"First 20: {sorted(list('nltk_stopwords'))[:20]}")

"First 20: ['_', 'd', 'k', 'l', 'n', 'o', 'o', 'p', 'r', 's', 's', 't', 't', 'w']"

In [72]:
spacy_stopwords = nlp.Defaults.stop_words

In [73]:
(f"Common stop words: {len('common_stopwords')}" "English stop words")

'Common stop words: 16English stop words'

In [74]:
(f"Only in spaCy: {len('spacy_only')} - Examples: {sorted(list('spacy_only'))[:5]}")

"Only in spaCy: 10 - Examples: ['_', 'a', 'c', 'l', 'n']"

In [75]:
'stopwords'

'stopwords'

In [76]:
nltk.corpus

<LazyModule 'nltk.corpus'>

### 🤔 Conceptual Question 7
**Why do you think NLTK and spaCy have different stop word lists? Look at the examples of words that are only in one list - do you agree with these choices? Can you think of scenarios where these differences might significantly impact your NLP results?**

*Double-click this cell to write your answer:*

**Reasons for differences:** The reasons why NLTK and spaCy have different stop word lists are because spaCy has 326 English stop words and NLTK has 198 Engligh stop words. NLTK has less Engligh stop words than spaCy. spaCy has more English stop words than NTLK.

**Agreement with choices:** I agree with this.

**Scenarios where differences matter:** NLTK tends to retain contractions and negations such as "I'm", "couldn't", "aren't", "he'd". spaCy gets rid off more contractions and auxiliary verbs, like: 'm, 'd, 'll, 're, 's. NLTK keeps contractions and negations. It's sentence structure is more intact. It's better for sentiment, but not for topic modeling. spaCy deletes contractions and negations. It's sentence structure is more fragment. It's not better for sentiment, but better for topic modeling.

Using NLTK involves my task being:
Sentiment analysis,
Text classification, and
Dialogue understanding.

Using spaCy involves my task being:
Topic modeling,
Keyword extraction, and
Search relevance.

---

In [77]:
# Step 8: Remove Stop Words with NLTK
# Test on simple text
original_tokens = nltk_tokens  # From earlier tokenization
filtered_tokens = [word for word in original_tokens if word.lower() not in nltk_stopwords]

print("🧪 NLTK Stop Word Removal")
print("=" * 40)
print(f"Original: {simple_text}")
print(f"\nOriginal tokens ({len(original_tokens)}): {original_tokens}")
print(f"After removing stop words ({len(filtered_tokens)}): {filtered_tokens}")

# Show which words were removed
removed_words = [word for word in original_tokens if word.lower() in nltk_stopwords]
print(f"\nRemoved words: {removed_words}")

# Calculate reduction percentage
reduction = (len(original_tokens) - len(filtered_tokens)) / len(original_tokens) * 100
print(f"Vocabulary reduction: {reduction:.1f}%")

🧪 NLTK Stop Word Removal
Original: Natural Language Processing is a fascinating field of AI. It's amazing!

Original tokens (14): ['Natural', 'Language', 'Processing', 'is', 'a', 'fascinating', 'field', 'of', 'AI', '.', 'It', "'s", 'amazing', '!']
After removing stop words (10): ['Natural', 'Language', 'Processing', 'fascinating', 'field', 'AI', '.', "'s", 'amazing', '!']

Removed words: ['is', 'a', 'of', 'It']
Vocabulary reduction: 28.6%


In [78]:
(f"\nOriginal tokens ({len(original_tokens)}): {original_tokens}")

'\nOriginal tokens (14): [\'Natural\', \'Language\', \'Processing\', \'is\', \'a\', \'fascinating\', \'field\', \'of\', \'AI\', \'.\', \'It\', "\'s", \'amazing\', \'!\']'

In [79]:
(f"Vocabualary reduction: {reduction:.1f}%")

'Vocabualary reduction: 28.6%'

In [80]:
reduction = (len('original_tokens') - len('filtered_tokens')) / len('original_tokens') * 100

In [81]:
(f"\nRemoved words: {removed_words}")

"\nRemoved words: ['is', 'a', 'of', 'It']"

In [82]:
removed_words = 'word for word in original_tokens if word.lower() in nltk_stopwords'

In [83]:
(f"After removing stop words ({len(filtered_tokens)}): {filtered_tokens}")

'After removing stop words (10): [\'Natural\', \'Language\', \'Processing\', \'fascinating\', \'field\', \'AI\', \'.\', "\'s", \'amazing\', \'!\']'

In [84]:
(f"Original tokens ({len('original_tokens')}): {filtered_tokens}")

'Original tokens (15): [\'Natural\', \'Language\', \'Processing\', \'fascinating\', \'field\', \'AI\', \'.\', "\'s", \'amazing\', \'!\']'

In [85]:
(f"Original: {simple_text}")

"Original: Natural Language Processing is a fascinating field of AI. It's amazing!"

In [86]:
(f"Original tokens ({len('original_tokens')}):")

'Original tokens (15):'

In [87]:
("= * 40")

'= * 40'

In [88]:
("🧪 NLTK Stop Word Removal")

'🧪 NLTK Stop Word Removal'

In [89]:
filtered_tokens = 'word for word in original_tokens if word.lower() not in nltk_stopwords'

In [90]:
original_tokens = 'nltk_tokens'

In [91]:
# Step 9: Remove Stop Words with spaCy
doc = nlp(simple_text)
spacy_filtered = [token.text for token in doc if not token.is_stop and not token.is_punct]

print("🧪 spaCy Stop Word Removal")
print("=" * 40)
print(f"Original: {simple_text}")
print(f"\nOriginal tokens ({len(spacy_tokens)}): {spacy_tokens}")
print(f"After removing stop words & punctuation ({len(spacy_filtered)}): {spacy_filtered}")

# Show which words were removed
spacy_removed = [token.text for token in doc if token.is_stop or token.is_punct]
print(f"\nRemoved words: {spacy_removed}")

# Calculate reduction percentage
spacy_reduction = (len(spacy_tokens) - len(spacy_filtered)) / len(spacy_tokens) * 100
print(f"Vocabulary reduction: {spacy_reduction:.1f}%")

🧪 spaCy Stop Word Removal
Original: Natural Language Processing is a fascinating field of AI. It's amazing!

Original tokens (14): ['Natural', 'Language', 'Processing', 'is', 'a', 'fascinating', 'field', 'of', 'AI', '.', 'It', "'s", 'amazing', '!']
After removing stop words & punctuation (7): ['Natural', 'Language', 'Processing', 'fascinating', 'field', 'AI', 'amazing']

Removed words: ['is', 'a', 'of', '.', 'It', "'s", '!']
Vocabulary reduction: 50.0%


In [92]:
(f"Vocubalary reduction: {spacy_reduction:.1f}%")

'Vocubalary reduction: 50.0%'

In [93]:
spacy_reduction = (len('spacy_tokens') - len('spacy_filtered')) / len('spacy_tokens') * 100

In [94]:
(f"\nRemoved words: {'spacy_removed'}")

'\nRemoved words: spacy_removed'

In [95]:
(f"Original: {simple_text}")

"Original: Natural Language Processing is a fascinating field of AI. It's amazing!"

In [96]:
(f"After removing stop words & punctuation ({len(spacy_filtered)}): {spacy_filtered}")

"After removing stop words & punctuation (7): ['Natural', 'Language', 'Processing', 'fascinating', 'field', 'AI', 'amazing']"

In [97]:
(f"\nOriginal tokens ({len(spacy_tokens)}): {spacy_tokens}")

'\nOriginal tokens (14): [\'Natural\', \'Language\', \'Processing\', \'is\', \'a\', \'fascinating\', \'field\', \'of\', \'AI\', \'.\', \'It\', "\'s", \'amazing\', \'!\']'

In [98]:
("=" * 40)

'========================================'

In [99]:
("🧪 spaCy Stop Word Removal")

'🧪 spaCy Stop Word Removal'

In [100]:
spacy_filtered = ['word for token in doc if not token.is_stop and not token.is_punct']

In [101]:
spacy_filtered = 'token.text for token in doc if not token.is_stop and not token.is_punct'

In [102]:
doc= nlp(simple_text)

### 🤔 Conceptual Question 8
**Compare the NLTK and spaCy stop word removal results. Which approach removed more words? Do you think removing punctuation (as spaCy did) is always a good idea? Give a specific example where keeping punctuation might be important for NLP analysis.**

*Double-click this cell to write your answer:*

**Which removed more:** NLTK removed more words than spaCy did.

**Punctuation removal assessment:** Removing punctuation is not always a good idea.

**Example where punctuation matters:**

Exaclaimation marks (!):
"I love this product!" vs "I love this product." - the exclaimation mark includes powerful positive sentiment.

Question marks (?): "Really?" can refer to doubt or sarcasm.

Ellipes (...) or multiple punctuation: "Well... I guess it's okay" could mean hesitation or uncertainty.

---

## 🌱 Part 5: Lemmatization and Stemming

### What is Lemmatization?
Lemmatization reduces words to their base or dictionary form (called a **lemma**). It considers context and part of speech to ensure the result is a valid word.

### What is Stemming?
Stemming reduces words to their root form by removing suffixes. It's faster but less accurate than lemmatization.

### Key Differences:
| Aspect | Stemming | Lemmatization |
|--------|----------|---------------|
| Speed | Fast | Slower |
| Accuracy | Lower | Higher |
| Output | May be non-words | Always valid words |
| Context | Ignores context | Considers context |

### Examples:
- **"running"** → Stem: "run", Lemma: "run"
- **"better"** → Stem: "better", Lemma: "good"
- **"was"** → Stem: "wa", Lemma: "be"

In [103]:
# Step 10: Stemming with NLTK
from nltk.stem import PorterStemmer

stemmer = PorterStemmer()

# Test words that demonstrate stemming challenges
test_words = ['running', 'runs', 'ran', 'better', 'good', 'best', 'flying', 'flies', 'was', 'were', 'cats', 'dogs']

print("🌿 Stemming Demonstration")
print("=" * 30)
print(f"{'Original':<12} {'Stemmed':<12}")
print("-" * 25)

for word in test_words:
    stemmed = stemmer.stem(word)
    print(f"{word:<12} {stemmed:<12}")

# Apply to our sample text
sample_tokens = [token for token in nltk_tokens if token.isalpha()]
stemmed_tokens = [stemmer.stem(token.lower()) for token in sample_tokens]

print(f"\n🧪 Applied to sample text:")
print(f"Original: {sample_tokens}")
print(f"Stemmed: {stemmed_tokens}")

🌿 Stemming Demonstration
Original     Stemmed     
-------------------------
running      run         
runs         run         
ran          ran         
better       better      
good         good        
best         best        
flying       fli         
flies        fli         
was          wa          
were         were        
cats         cat         
dogs         dog         

🧪 Applied to sample text:
Original: ['Natural', 'Language', 'Processing', 'is', 'a', 'fascinating', 'field', 'of', 'AI', 'It', 'amazing']
Stemmed: ['natur', 'languag', 'process', 'is', 'a', 'fascin', 'field', 'of', 'ai', 'it', 'amaz']


In [104]:
("-" * 25)

'-------------------------'

In [105]:
(f"{'Original':<12} {'Stemmed':<12}")

'Original     Stemmed     '

In [106]:
("=" * 30)

'=============================='

In [107]:
("🌿 Stemming Demonstration")

'🌿 Stemming Demonstration'

In [108]:
stemmed = 'stemmer.stem(word)'

In [109]:
(f"{'word':<12} {'stemmed':<12}")

'word         stemmed     '

In [110]:
from random import sample
sample_tokens = 'token for token in nltk_tokens if token.isalpha()'

In [111]:
(f"Stemmed: {'stemmed_tokens'}")

'Stemmed: stemmed_tokens'

In [112]:
(f"Original: {'sample_tokens'}")

'Original: sample_tokens'

In [113]:
(f"\n🧪 Applied to sample text:")

'\n🧪 Applied to sample text:'

In [114]:
test_words = ['running', 'runs', 'ran', 'better', 'good', 'best', 'flying', 'flies', 'was', 'were', 'cats', 'dogs']

In [115]:
test_words = 'running', 'runs', 'ran', 'better', 'good', 'best', 'flying', 'flies', 'was', 'were', 'cats', 'dogs'

In [116]:
stemmer = 'PorterStemmer'

In [117]:
nltk.stem

<module 'nltk.stem' from '/usr/local/lib/python3.12/dist-packages/nltk/stem/__init__.py'>

In [118]:
'PorterStemmer'

'PorterStemmer'

### 🤔 Conceptual Question 9
**Look at the stemming results above. Can you identify any cases where stemming produced questionable results? For example, how were "better" and "good" handled? Do you think this is problematic for NLP applications? Explain your reasoning.**

*Double-click this cell to write your answer:*

**Questionable results identified:**
1. "better" - better

2. "good" - good

3. "best" - best

4. "flying" - fli, "flies" - fli

5. "was" - wa



**Assessment of "better" and "good":** Stemming can occassionally create questionable results that impact the quality of NLP applications. For the examples, words like "better", "good", and "best" were not decreased to a common root, even though they're particulary relevant.

**Impact on NLP applications:** The 3 words not being decreased to a common root will trigger fragmented meaning in tasks like sentiment analysis or text classification. Addtionally, overly aggressive stemming-like "flying" and "flies" becoming "fli", or "was" becoming "wa"-results in non-words that decrease interpretability and may complicate models.

---

In [119]:
# Step 11: Lemmatization with spaCy
print("🌱 spaCy Lemmatization Demonstration")
print("=" * 40)

# Test on a complex sentence
complex_sentence = "The researchers were studying the effects of running and swimming on better performance."
doc = nlp(complex_sentence)

print(f"Original: {complex_sentence}")
print(f"\n{'Token':<15} {'Lemma':<15} {'POS':<10} {'Explanation':<20}")
print("-" * 65)

for token in doc:
    if token.is_alpha:
        explanation = "No change" if token.text.lower() == token.lemma_ else "Lemmatized"
        print(f"{token.text:<15} {token.lemma_:<15} {token.pos_:<10} {explanation:<20}")

# Extract lemmas
lemmas = [token.lemma_.lower() for token in doc if token.is_alpha and not token.is_stop]
print(f"\n🔤 Lemmatized tokens (no stop words): {lemmas}")

🌱 spaCy Lemmatization Demonstration
Original: The researchers were studying the effects of running and swimming on better performance.

Token           Lemma           POS        Explanation         
-----------------------------------------------------------------
The             the             DET        No change           
researchers     researcher      NOUN       Lemmatized          
were            be              AUX        Lemmatized          
studying        study           VERB       Lemmatized          
the             the             DET        No change           
effects         effect          NOUN       Lemmatized          
of              of              ADP        No change           
running         run             VERB       Lemmatized          
and             and             CCONJ      No change           
swimming        swim            VERB       Lemmatized          
on              on              ADP        No change           
better          well          

In [120]:
(f"\n Lemmatized tokens (no stop words): {'lemmas'}")

'\n Lemmatized tokens (no stop words): lemmas'

In [121]:
lemmas = 'token.lemma_.lower() for token in doc if token.is_alpha and not token.is_stop'

In [122]:
(f"{token.text:<15} {token.lemma_:<15} {token.pos_:<10} {'explanation':<20}")

'.               .               PUNCT      explanation         '

In [123]:
if token.is_alpha: explanation = "No change" if token.text.lower() == token.lemma_ else "Lemmatized"

In [124]:
("-" * 65)

'-----------------------------------------------------------------'

In [125]:
(f"Original: {complex_sentence}")

'Original: The researchers were studying the effects of running and swimming on better performance.'

In [126]:
doc = nlp(complex_sentence)

In [127]:
complex_sentence = 'The researchers were studying the effects of running and swimming on better performance.'

In [128]:
("=" * 40)

'========================================'

In [129]:
(" spaCy Lemmatization Demonstration")

' spaCy Lemmatization Demonstration'

In [130]:
(f"\n{'Token':<15} {'Lemma':<15} {'POS':<10} {'Explanation':<20}")

'\nToken           Lemma           POS        Explanation         '

In [147]:
import nltk
from nltk.stem import PorterStemmer

# Step 12: Compare Stemming vs Lemmatization
comparison_words = ['better', 'running', 'studies', 'was', 'children', 'feet']

print("⚖️ Stemming vs Lemmatization Comparison")
print("=" * 50)
print(f"{'Original':<12} {'Stemmed':<12} {'Lemmatized':<12}")
print("-" * 40)

# Ensure stemmer is correctly initialized as an object
stemmer = PorterStemmer()

for word in comparison_words:
    # Stemming
    stemmed = stemmer.stem(word)

    # Lemmatization with spaCy
    doc = nlp(word)
    lemmatized = doc[0].lemma_

    print(f"{word:<12} {stemmed:<12} {lemmatized:<12}")

⚖️ Stemming vs Lemmatization Comparison
Original     Stemmed      Lemmatized  
----------------------------------------
better       better       well        
running      run          run         
studies      studi        study       
was          wa           be          
children     children     child       
feet         feet         foot        


In [143]:
(f"{word:<12} {stemmed:<12} {lemmatized:<12}")

'better       stemmer.stem(word) doc[0].lemma_'

In [145]:
for word in comparison_words:
    stemmed = 'stemmer.stem(word)'
    doc = nlp(word)
    lemmatized = 'doc[0].lemma_'

In [135]:
("Stemming vs Lemmatization Comparison")

'Stemming vs Lemmatization Comparison'

In [136]:
("-" * 40)

'----------------------------------------'

In [137]:
("=" * 50)

'=================================================='

In [138]:
(f"\n{'Original':<12} {'Stemmed':<12} {'Lemmatized':<12}")

'\nOriginal     Stemmed      Lemmatized  '

In [139]:
comparison_words = 'better', 'running', 'studies', 'was', 'children', 'feet'

### 🤔 Conceptual Question 10
**Compare the stemming and lemmatization results. Which approach do you think is more suitable for:**
1. **A search engine** (where speed is crucial and you need to match variations of words)?
2. **A sentiment analysis system** (where accuracy and meaning preservation are important)?
3. **A real-time chatbot** (where both speed and accuracy matter)?

**Explain your reasoning for each choice.**

*Double-click this cell to write your answer:*

**1. Search engine:** The search engine would likely use Stemming. Stemming rapidly decreases words to their root forms, which helps to match opposite word variant in user queries and documents. However it might create non-words like "fli" or "wa", this is generally acceptable in search engines where memorization is prioritized over perfect linguistic accuracy.

**2. Sentiment analysis:** Sentiment analysis would be best used for Lemmatization. Lemmatization maintains the semantic meaning of the words better than stemming. For example, "better", "good", and "best" are handled more smartly, with "better" and "good" both being decreased to "good". This helps the system learn sentiment intensity and relationships among words, which significant for correct sentiment scoring.

**3. Real-time chatbot:** The real-time chatbot would be best used for lemmatization. Chatbots need to learn and answer with natural language. Lemmatization stays away from creating unnatural or non-words such as "fli" or "wa", which stemming may produce. By keeping words meaningful and grammatically accurate, lemmatization supports better language understanding and generation, enhancing user experience.

---

## 🧹 Part 6: Text Cleaning and Normalization

### What is Text Cleaning?
Text cleaning involves removing or standardizing elements that might interfere with analysis:
- **Case normalization** (converting to lowercase)
- **Punctuation removal**
- **Number handling** (remove, replace, or normalize)
- **Special character handling** (URLs, emails, mentions)
- **Whitespace normalization**

### Why is it Important?
- Ensures consistency across your dataset
- Reduces vocabulary size
- Improves model performance
- Handles edge cases in real-world data

In [148]:
# Step 13: Basic Text Cleaning
def basic_clean_text(text):
    """Apply basic text cleaning operations"""
    # Convert to lowercase
    text = text.lower()

    # Remove extra whitespace
    text = re.sub(r'\s+', ' ', text).strip()

    # Remove punctuation
    text = text.translate(str.maketrans('', '', string.punctuation))

    # Remove numbers
    text = re.sub(r'\d+', '', text)

    # Remove extra spaces again
    text = re.sub(r'\s+', ' ', text).strip()

    return text

# Test basic cleaning
test_text = "   Hello WORLD!!! This has 123 numbers and   extra spaces.   "
cleaned = basic_clean_text(test_text)

print("🧹 Basic Text Cleaning")
print("=" * 30)
print(f"Original: '{test_text}'")
print(f"Cleaned: '{cleaned}'")
print(f"Length reduction: {(len(test_text) - len(cleaned))/len(test_text)*100:.1f}%")

🧹 Basic Text Cleaning
Original: '   Hello WORLD!!! This has 123 numbers and   extra spaces.   '
Cleaned: 'hello world this has numbers and extra spaces'
Length reduction: 26.2%


In [149]:
(f"Length reduction: {(len(test_text) - len(cleaned))/len(test_text)*100:.1f}%")

'Length reduction: 26.2%'

In [150]:
(f"Original: '{test_text}'")

"Original: '   Hello WORLD!!! This has 123 numbers and   extra spaces.   '"

In [151]:
("=" * 30)

'=============================='

In [152]:
(" Basic Text Cleaning")

' Basic Text Cleaning'

In [153]:
(f"Cleaned: '{cleaned}'")

"Cleaned: 'hello world this has numbers and extra spaces'"

In [154]:
cleaned = basic_clean_text(test_text)

In [155]:
test_text = '   Hello WORLD!!! This has 123 numbers and   extra spaces.   '

In [156]:
def basic_clean_text(test_text):
    """Apply basic text cleaning operations"""
    # Convert to lowercase
    test_text = test_text.lower()

    # Remove extra whitespace
    test_text = re.sub(r'\s+', ' ', test_text).strip()

    # Remove punctuation
    test_text = test_text.translate(str.maketrans('', '', string.punctuation))

    # Remove numbers
    test_text = re.sub(r'\d+', '', test_text)

    # Remove extra spaces
    test_text = re.sub(r'\s+', ' ', test_text).strip()

    return test_text

In [199]:
def advanced_clean_text(text):
    """Apply advanced cleaning for social media and web text"""
    # Remove URLs
    text = re.sub(r'http\S+|www\S+|https\S+', '', text, flags=re.MULTILINE)

    # Remove email addresses
    text = re.sub(r'\S+@\S+', '', text)

    # Remove mentions (@username)
    text = re.sub(r'@\w+', '', text)

    # Convert hashtags (keep the word, remove #)
    text = re.sub(r'#(\w+)', r'\1', text)

    # Remove emojis (basic approach)
    emoji_pattern = re.compile("["
                               u"\U0001F600-\U0001F64F"  # emoticons
                               u"\U0001F300-\U0001F5FF"  # symbols & pictographs
                               u"\U0001F680-\U0001F6FF"  # transport & map symbols
                               u"\U0001F1E0-\U0001F1FF"  # flags
                               "]+", flags=re.UNICODE)
    text = emoji_pattern.sub(r'', text)

    # Convert to lowercase and normalize whitespace
    text = text.lower()
    text = re.sub(r'\s+', ' ', text).strip()

    return text

# Test on social media text
print("🚀 Advanced Cleaning on Social Media Text")
print("=" * 45)
print(f"Original: {social_text}")

cleaned_social = advanced_clean_text(social_text)
print(f"Cleaned: {cleaned_social}")
print(f"Length reduction: {(len(social_text) - len(cleaned_social))/len(social_text)*100:.1f}%")

🚀 Advanced Cleaning on Social Media Text
Original: OMG! Just tried the new coffee shop ☕️ SO GOOD!!! Highly recommend 👍 #coffee #yum 😍
Cleaned: omg! just tried the new coffee shop ☕️ so good!!! highly recommend coffee yum
Length reduction: 7.2%


In [158]:
def advanced_clean_text(social_text):
    """Apply advanced cleaning for social media and web text"""
    # Remove URLs
    text = re.sub(r'http\S+|www\S+|https\S+', '', text, flags=re.MULTILINE)

    # Remove email addresses
    text = re.sub(r'\S+@\S+', '', text)

In [159]:
#Remove mentions (@username)
text = re.sub(r'@\w+', '', text)

In [160]:
# Convert hastags (keep the word, remove #)
text = re.sub(r'#(\w+)', r'\1', text)

In [161]:
# Remove emojis (basic approach)
emoji_pattern = re.compile("["
u"\U0001F600-\U0001F64F"
u"\U0001F300-\U0001F5FF"
u"\U0001F680-\U0001F6FF"
u"\U0001F1E0-\U0001F1FF"
"]+", flags=re.UNICODE)

In [162]:
text = emoji_pattern.sub(r'', text)

In [163]:
# Covert to lowercase and normalize whitespace
text = text.lower()
text = re.sub(r'\s+', ' ', text).strip()
'return text'

'return text'

In [164]:
# Test on social media text
print("🚀 Advanced Cleaning on Social Media Text")
print("=" * 45)
print(f"Original: {social_text}")

🚀 Advanced Cleaning on Social Media Text
Original: OMG! Just tried the new coffee shop ☕️ SO GOOD!!! Highly recommend 👍 #coffee #yum 😍


In [165]:
cleaned_social = 'advanced_clean_text(social_text)'

In [166]:
(f"Cleaned: {cleaned_social}")

'Cleaned: advanced_clean_text(social_text)'

In [167]:
(f"Length reduction: {(len(social_text) - len(cleaned_social))/len(social_text)*100:.1f}%")

'Length reduction: 61.4%'

### 🤔 Conceptual Question 11
**Look at the advanced cleaning results for the social media text. What information was lost during cleaning? Can you think of scenarios where removing emojis and hashtags might actually hurt your NLP application? What about scenarios where keeping them would be beneficial?**

*Double-click this cell to write your answer:*

**Information lost:** In the advanced cleaning process, emojis and hastags like #coffee and #yum were removed. This triggered to a loss of valuable emotional and topical details.

**Scenarios where removal hurts:** Deleting the elements could hurt NLP applications like sentiment analysis, where emojis express powerful feelings, and topic modeling, where hastags help spot themes.

**Scenarios where keeping helps:** Keeping emojis and hastags is helpful in social media monitoring and customer feedback analysis, where they provide penetrations into user engagement, trending subjects, and emotional tone. Securing them benefits the richness and context of informal online communication.

---

## 🔧 Part 7: Building a Complete Preprocessing Pipeline

Now let's combine everything into a comprehensive preprocessing pipeline that you can customize based on your needs.

### Pipeline Components:
1. **Text cleaning** (basic or advanced)
2. **Tokenization** (NLTK or spaCy)
3. **Stop word removal** (optional)
4. **Lemmatization/Stemming** (optional)
5. **Additional filtering** (length, etc.)

In [168]:
# Step 15: Complete Preprocessing Pipeline
def preprocess_text(text,
                   clean_level='basic',     # 'basic' or 'advanced'
                   remove_stopwords=True,
                   use_lemmatization=True,
                   use_stemming=False,
                   min_length=2):
    """
    Complete text preprocessing pipeline
    """
    # Step 1: Clean text
    if clean_level == 'basic':
        cleaned_text = basic_clean_text(text)
    else:
        cleaned_text = advanced_clean_text(text)

    # Step 2: Tokenize
    if use_lemmatization:
        # Use spaCy for lemmatization
        doc = nlp(cleaned_text)
        tokens = [token.lemma_.lower() for token in doc if token.is_alpha]
    else:
        # Use NLTK for basic tokenization
        tokens = word_tokenize(cleaned_text)
        tokens = [token for token in tokens if token.isalpha()]

    # Step 3: Remove stop words
    if remove_stopwords:
        if use_lemmatization:
            tokens = [token for token in tokens if token not in spacy_stopwords]
        else:
            tokens = [token.lower() for token in tokens if token.lower() not in nltk_stopwords]

    # Step 4: Apply stemming if requested
    if use_stemming and not use_lemmatization:
        tokens = [stemmer.stem(token.lower()) for token in tokens]

    # Step 5: Filter by length
    tokens = [token for token in tokens if len(token) >= min_length]

    return tokens

print("🔧 Preprocessing Pipeline Created!")
print("✅ Ready to test different configurations.")

🔧 Preprocessing Pipeline Created!
✅ Ready to test different configurations.


In [169]:
if 'clean_level' == 'basic':
    cleaned_text = basic_clean_text(text)

In [170]:
cleaned_text = 'basic_clean_text(text)'

In [171]:
'use_lemmatization'

'use_lemmatization'

In [172]:
doc = 'nlp'

In [173]:
doc = (cleaned_text)

In [174]:
tokens = ['token.lemma_.lower() for token in doc if token.is_alpha']

In [175]:
tokens = [token for token in tokens if token.isalpha()]

In [176]:
tokens = [token for token in tokens if token not in spacy_stopwords]

In [177]:
tokens = [token.lower() for token in tokens if token.lower() not in nltk_stopwords]

In [178]:
tokens = [stemmer.stem(token.lower()) for token in tokens]

In [179]:
tokens = [token for token in tokens if len(token) >= min_length]

In [180]:
'return tokens'
("🔧 Preprocessing Pipeline Created!")
("✅ Ready to test different configurations.")

'✅ Ready to test different configurations.'

In [200]:
# Step 16: Test Different Pipeline Configurations
test_text = sample_texts["Product Review"]
print(f"🎯 Testing on: {test_text[:100]}...")
print("=" * 60)

# Configuration 1: Minimal processing
minimal = preprocess_text(test_text,
                         clean_level='basic',
                         remove_stopwords=False,
                         use_lemmatization=False,
                         use_stemming=False)
print(f"\n1. Minimal processing ({len(minimal)} tokens):")
print(f"   {minimal[:10]}...")

# Configuration 2: Standard processing
standard = preprocess_text(test_text,
                          clean_level='advanced', # Changed from True to 'advanced'
                          remove_stopwords=True,
                          use_lemmatization=True)
print(f"\n2. Standard processing ({len(standard)} tokens):")
print(f"   {standard[:10]}...")

# Configuration 3: Aggressive processing
aggressive = preprocess_text(test_text,
                            clean_level='advanced',
                            remove_stopwords=True,
                            use_lemmatization=False,
                            use_stemming=True,
                            min_length=3)
print(f"\n3. Aggressive processing ({len(aggressive)} tokens):")
print(f"   {aggressive[:10]}...")

# Show reduction percentages
original_count = len(word_tokenize(test_text))
print(f"\n📊 Token Reduction Summary:")
print(f"   Original: {original_count} tokens")
print(f"   Minimal: {len(minimal)} ({(original_count-len(minimal))/original_count*100:.1f}% reduction)")
print(f"   Standard: {len(standard)} ({(original_count-len(standard))/original_count*100:.1f}% reduction)")
print(f"   Aggressive: {len(aggressive)} ({(original_count-len(aggressive))/original_count*100:.1f}% reduction)")

🎯 Testing on: This laptop is absolutely fantastic! I've been using it for 6 months and it's still super fast.
The ...

1. Minimal processing (34 tokens):
   ['this', 'laptop', 'is', 'absolutely', 'fantastic', 'ive', 'been', 'using', 'it', 'for']...

2. Standard processing (18 tokens):
   ['laptop', 'absolutely', 'fantastic', 'use', 'month', 'super', 'fast', 'battery', 'life', 'incredible']...

3. Aggressive processing (21 tokens):
   ['laptop', 'absolut', 'fantast', 'use', 'month', 'still', 'super', 'fast', 'batteri', 'life']...

📊 Token Reduction Summary:
   Original: 47 tokens
   Minimal: 34 (27.7% reduction)
   Standard: 18 (61.7% reduction)
   Aggressive: 21 (55.3% reduction)


In [182]:
nltk.download('punkt_tab')

[nltk_data] Downloading package punkt_tab to /root/nltk_data...
[nltk_data]   Package punkt_tab is already up-to-date!


True

In [183]:
(f"\n1. Minimal processing ({len('minimal')} tokens)")

'\n1. Minimal processing (7 tokens)'

In [185]:
(f"{'minimal' [:10]}...")

'minimal...'

In [186]:
(f"\n2. Standard processing ({len('standard')} tokens)")

'\n2. Standard processing (8 tokens)'

In [187]:
(f"{'standard' [:10]}...")

'standard...'

In [188]:
(f"\n3. Aggressive processing ({len('aggressive')} tokens)")

'\n3. Aggressive processing (10 tokens)'

In [189]:
(f"{'aggressive' [:10]}...")

'aggressive...'

In [190]:
(f"\n Token Reduction Summary:")

'\n Token Reduction Summary:'

In [202]:
(f" Original: {original_count} tokens")

' Original: 47 tokens'

In [203]:
(f" Minimal: {len('minimal')} ({(original_count-len('minimal'))/original_count*100:.1f}% reduction")

' Minimal: 7 (85.1% reduction'

In [204]:
print(f" Standard: {len('standard')} ({(original_count-len('standard'))/original_count*100:.1f}% reduction)")

 Standard: 8 (83.0% reduction)


In [196]:
(f"🎯 Testing on: {test_text[:100]}...")

"🎯 Testing on: This laptop is absolutely fantastic! I've been using it for 6 months and it's still super fast.\nThe ..."

In [205]:
(f"   Aggressive: {len('aggressive')} ({(original_count-len('aggressive'))/original_count*100:.1f}% reduction)")

'   Aggressive: 10 (78.7% reduction)'

In [206]:
("=" * 60)

'============================================================'

In [207]:
'sample_texts'

'sample_texts'

In [208]:
(f"🎯 Testing on: {test_text[:100]}...")

"🎯 Testing on: This laptop is absolutely fantastic! I've been using it for 6 months and it's still super fast.\nThe ..."

In [209]:
("=" * 60)

'============================================================'

In [210]:
(f"\n1. Minimal processing ({len('minimal')} tokens):")

'\n1. Minimal processing (7 tokens):'

In [211]:
(f" {'minimal'[:10]}...")

' minimal...'

In [212]:
(f"\n2. Standard processing ({len('standard')} tokens):")

'\n2. Standard processing (8 tokens):'

In [213]:
(f" {'standard'[:10]}...")

' standard...'

In [214]:
(f"\n3. Aggressive processing ({len('aggressive')} tokens):")

'\n3. Aggressive processing (10 tokens):'

In [215]:
(f" {'aggressive'[:10]}...")

' aggressive...'

In [216]:
(f"\n3. Aggressive processing ({len('aggressive')} tokens):")

'\n3. Aggressive processing (10 tokens):'

In [217]:
(f" {'aggressive'[:10]}...")

' aggressive...'

In [218]:
original_count = len('word_tokenize')

In [219]:
original_count = len(test_text)

In [220]:
(f" Original count: {original_count} tokens")

' Original count: 228 tokens'

In [221]:
(f" Minimal: {len('minimal')} ({(original_count-len('minimal'))/original_count*100:.1f}% reduction)")

' Minimal: 7 (96.9% reduction)'

In [222]:
(f" Standard: {len('standard')} ({(original_count-len('standard'))/original_count*100:.1f}% reduction)")

' Standard: 8 (96.5% reduction)'

In [223]:
(f" Aggressive: {len('aggressive')} ({(original_count-len('aggressive'))/original_count*100:.1f}% reduction)")

' Aggressive: 10 (95.6% reduction)'

### 🤔 Conceptual Question 12
**Compare the three pipeline configurations (Minimal, Standard, Aggressive). For each configuration, analyze:**
1. **What information was preserved?**
2. **What information was lost?**
3. **What type of NLP task would this configuration be best suited for?**

*Double-click this cell to write your answer:*

**Minimal Processing:**
- Preserved: Core words and basic sentence structure.
- Lost: Captilization, punctation, numbers, and extra spaces.
- Best for: Keyword extraction, Basic sentiment analysis, and Text normalization for search.

**Standard Processing:**
- Preserved: Core words, some punctuation, emojis, and hashtags (transformed to plain text)
- Lost: Redundant emojis, several unique characters.
- Best for: Social media monitoring, Opinion mining, and Customer feedback analysis.

**Aggressive Processing:**
- Preserved: Just important words
- Lost: Punctuation, emojis, hashtags, and stylistic elements.
- Best for: Topic modeling, Large-scale summarization, and Corpus-level analysis.

---

In [201]:
# Step 17: Comprehensive Analysis Across Text Types
print("🔬 Comprehensive Preprocessing Analysis")
print("=" * 50)

# Test standard preprocessing on all text types
results = {}
for name, text in sample_texts.items():
    original_tokens = len(word_tokenize(text))
    processed_tokens = preprocess_text(text,
                                      clean_level='basic',
                                      remove_stopwords=True,
                                      use_lemmatization=True)

    reduction = (original_tokens - len(processed_tokens)) / original_tokens * 100
    results[name] = {
        'original': original_tokens,
        'processed': len(processed_tokens),
        'reduction': reduction,
        'sample': processed_tokens[:8]
    }

    print(f"\n📄 {name}:")
    print(f"   Original: {original_tokens} tokens")
    print(f"   Processed: {len(processed_tokens)} tokens ({reduction:.1f}% reduction)")
    print(f"   Sample: {processed_tokens[:8]}")

# Summary table
print(f"\n\n📋 Summary Table")
print(f"{'Text Type':<15} {'Original':<10} {'Processed':<10} {'Reduction':<10}")
print("-" * 50)
for name, data in results.items():
    print(f"{name:<15} {data['original']:<10} {data['processed']:<10} {data['reduction']:<10.1f}%")

🔬 Comprehensive Preprocessing Analysis

📄 Simple:
   Original: 14 tokens
   Processed: 7 tokens (50.0% reduction)
   Sample: ['natural', 'language', 'processing', 'fascinating', 'field', 'ai', 'amazing']

📄 Academic:
   Original: 61 tokens
   Processed: 26 tokens (57.4% reduction)
   Sample: ['dr', 'smith', 'research', 'machinelearning', 'algorithm', 'groundbreake', 'publish', 'paper']

📄 Social Media:
   Original: 22 tokens
   Processed: 10 tokens (54.5% reduction)
   Sample: ['omg', 'try', 'new', 'coffee', 'shop', 'good', 'highly', 'recommend']

📄 News:
   Original: 51 tokens
   Processed: 25 tokens (51.0% reduction)
   Sample: ['stock', 'market', 'experience', 'significant', 'volatility', 'today', 'tech', 'stock']

📄 Product Review:
   Original: 47 tokens
   Processed: 18 tokens (61.7% reduction)
   Sample: ['laptop', 'absolutely', 'fantastic', 've', 'use', 'month', 'super', 'fast']


📋 Summary Table
Text Type       Original   Processed  Reduction 
----------------------------------

In [224]:
reduction = (original_tokens - len(processed_tokens)) / original_tokens * 100

In [225]:
processed_tokens = preprocess_text(text,
                                  clean_level='basic',
                                  remove_stopwords=True,
                                  use_lemmatization=True)

In [226]:
original_tokens = len(word_tokenize(text))

In [227]:
(f"\n {name}:")

'\n Product Review:'

In [228]:
(f"{'name':<15} {'data[original]:<10'} {'data[processed]:<10'} {'data[reduction]:<10.1f'}%")

'name            data[original]:<10 data[processed]:<10 data[reduction]:<10.1f%'

In [229]:
("-" * 50)

'--------------------------------------------------'

In [230]:
(f"{'Text Type':<15} {'Original':<10} {'Processed':<10} {'Reduction':<10}")

'Text Type       Original   Processed  Reduction '

In [231]:
("=" * 50)

'=================================================='

In [232]:
(f"  Sample: {'processed_tokens'[:8]}")

'  Sample: processe'

In [233]:
("   Processed: {len('processed_tokens')} tokens ({'reduction':.1f}% 'reduction')")

"   Processed: {len('processed_tokens')} tokens ({'reduction':.1f}% 'reduction')"

In [234]:
(f"  Original: {'original_tokens'} tokens")

'  Original: original_tokens tokens'

In [235]:
(f"Step 17: Comprehensive Analysis Across Text Types")

'Step 17: Comprehensive Analysis Across Text Types'

In [236]:
(f"🔬 Comprehensive Preprocessing Analysis")

'🔬 Comprehensive Preprocessing Analysis'

In [237]:
(f"\n\n📋 Summary Table")

'\n\n📋 Summary Table'

### 🤔 Final Conceptual Question 13
**Looking at the comprehensive analysis results across all text types:**

1. **Which text type was most affected by preprocessing?** Why do you think this happened?

2. **Which text type was least affected?** What does this tell you about the nature of that text?

3. **If you were building an NLP system to analyze customer reviews for a business, which preprocessing approach would you choose and why?**

4. **What are the main trade-offs you need to consider when choosing preprocessing techniques for any NLP project?**

*Double-click this cell to write your answer:*

**1. Most affected text type:** The Product Review text type was the most affected by preprocessing, with a 61.7% reduction in token count. This likely occurred because product reviews usually have informal langauge, repetitve phrases, emphasis words, and non-standard formatting, which are stripped when cleaning. The expressive components are valuable for sentiment analysis, so agressive cleaning might risk losing emotional nuance.

**2. Least affected text type:** The Simple text type was the least affected, with only a 50.0% reduction. This reccomends that simple texts are already neat and possess fewer noisy or redundant components. Their structure is simple, making them less unsafe to preprocessing loss.

**3. For customer review analysis:** If I'm building an NLP system to examine customer reviews, I would possibly choose a standard preprocessing approach that removes sound (extra spaces, irrelevant punctuation, and stop words). One that maintains meaningful content like emoji, negations, and domain-specific terms. The two reasons why I would choose that are because customer reviews particulary have emotional cues and informal expressions that are significant for learning sentiment. Over-agressive cleaning can strip away valuabe context and decrease correction in tasks such as sentiment analysis, aspect-based opinion mining, or intent detection.

**4. Main trade-offs to consider:**
1. Cleanlieness vs. Context -
More cleaning = less sound, but even less nuance. Deleting emojis, hashtags, or contractiosns might clarify text but lose emotional or topical signals.

2. Speed vs. Accuracy -
Straightfoward methods (such as stemming) are quick but might distort meaning. More advanced techniques (such as lemmatization or contextula normalization) are slower but maintian semantics better.

3. Generalization vs. Specificity -
Generic cleaning works across various areas but might miss domain specific patterns. Custom cleaning modified to my dataset enhances relevance.

4. Interpretability vs. Model Performance -
Clean, structured text is simpler to explain. Raw or suggestive text could increase model performance in deep learning but be complicated to explain.

---

## 🎯 Lab Summary and Reflection

Congratulations! You've completed a comprehensive exploration of NLP preprocessing techniques.

### 🔑 Key Concepts You've Mastered:

1. **Text Preprocessing Fundamentals** - Understanding why preprocessing is crucial
2. **Tokenization Techniques** - NLTK vs spaCy approaches and their trade-offs
3. **Stop Word Management** - When to remove them and when to keep them
4. **Morphological Processing** - Stemming vs lemmatization for different use cases
5. **Text Cleaning Strategies** - Basic vs advanced cleaning for different text types
6. **Pipeline Design** - Building modular, configurable preprocessing systems

### 🎓 Real-World Applications:
These techniques form the foundation for search engines, chatbots, sentiment analysis, document classification, machine translation, and information extraction systems.

### 💡 Key Insights to Remember:
- **No Universal Solution**: Different NLP tasks require different preprocessing approaches
- **Trade-offs Are Everywhere**: Balance information preservation with noise reduction
- **Context Matters**: The same technique can help or hurt depending on your use case
- **Experimentation Is Key**: Always test and measure impact on your specific task

---

**Excellent work completing Lab 02!** 🎉

For your reflection journal, focus on the insights you gained about when and why to use different techniques, the challenges you encountered, and connections you made to real-world applications.